### PR TITLE
[MM-12852] E2E: Write initial test setup with Cypress

### PR DESCRIPTION
**Pull request by saturninoabril**
#### Summary
Write initial test setup with Cypress
- added default folder structure
```
-- cypress
   |-- fixtures // test data
   |-- integration // test cases 
   |-- support // commands, helpers
```
- added script at `package.json` such as 
  - `cypress:run` to run all tests via CLI with Cypress
  - `cypress:open` to open desktop app of Cypress
- added dev dependencies such as `cypress@3.1.1` and `eslint-plugin-cypress@2.1.2` 
- modified ESLint config with additional plugin and global setup for Cypress

Note: All items related to NightwatchJS are intentionally remain as is.

#### Ticket Link
Jira ticket: [MM-12852](https://mattermost.atlassian.net/browse/MM-12852)

#### Checklist
- [x] Ran `make check-style` to check for style errors (required for all pull requests)
- [x] Ran `make test` to ensure unit and component tests passed

### CODE REVIEW

File: components/login/login_controller/login_controller.jsx [hunk 1]
```diff
@@ -553,17 +553,18 @@ class LoginController extends React.Component {
...
-                        className='fa fa-refresh icon--rotate'
-                        title={Utils.localizeMessage('generic_icons.loading', 'Loading Icon')}
-                    />
-                    <FormattedMessage
-                        id='login.signInLoading'
-                        defaultMessage='Signing in...'
-                    />
-                </span>);
+                loginButton = (
+                    <span data-cy='login_signing_in'>
```
Comment (enahum) [inline components/login/login_controller/login_controller.jsx]: do we need to add `data-cy` to everything that we want to test? can you explain this a bit?

Comment (saturninoabril) [reply components/login/login_controller/login_controller.jsx]: Ref: https://docs.cypress.io/guides/references/best-practices.html#Selecting-Elements

It is the recommended best practice in selecting an element.  I'm particularly concern about "selectors break from development changes" which I think we'll definitely hit us from time to time.  

For most cases, I think we're good at using element ID but for us to be more careful on the change we'll be making, seeing `data-*` will mean that such element is important and is being *watched*.
Most likely, we'll be adding `data-*` to the following:
- a page, to easily identify if it has been loaded
- a section, like RHS, modal, etc
- a dynamic component/element
- an element being used in transition
- an element that addresses or has history of bug

**Comment by jwilander** [comment]
I know we didn't get a chance to see your demo during the dev meeting today because we ran out of time but I'd really like to see one or learn something about cypress before I approve

File: cypress/integration/login/login_a_elem_spec.js [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+        cy.get('.signup-team__container').should('be.visible');
+        cy.get('h1').should('contain', 'Mattermost');
+        cy.get('.color--light > span').should('contain', 'All team communication in one place, searchable and accessible anywhere');
+        cy.get('#loginId').should('be.visible');
+        cy.get('#loginId').should('have.attr', 'placeholder', 'Email or Username');
+        cy.get('#loginPassword').should('be.visible');
+        cy.get('#loginPassword').should('have.attr', 'placeholder', 'Password');
+        cy.get('#loginButton').should('be.visible');
+        cy.get('#loginButton > span').should('contain', 'Sign in');
+        cy.get(':nth-child(3) > :nth-child(2) > a > span').should('contain', 'I forgot my password');
```
Comment (jwilander) [inline cypress/integration/login/login_a_elem_spec.js]: As discussed, don't use children to find an element

File: cypress/integration/login/login_b_action_spec.js [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+        // 4. Click "Sign in" button
+        cy.get('#loginButton').click();
+
+        // * Check that the Signin button change with rotating icon and "Signing in..." text
+        cy.get('[data-cy="login_signing_in"]').should('be.visible');
+        cy.get('#loginButton > :nth-child(1) > :nth-child(2)').should('contain', 'Signing in...');
+
+        // * Check that it login successfully and it redirects into the main channel page
+        cy.get('[data-cy="main_channel_page"]').should('be.visible');
```
Comment (jwilander) [inline cypress/integration/login/login_b_action_spec.js]: As discussed, can we use ids instead

File: cypress/integration/login/login_b_action_spec.js [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+        // 5. Click hamburger main menu button
+        cy.get('.sidebar-header-dropdown__icon').click();
```
Comment (jwilander) [inline cypress/integration/login/login_b_action_spec.js]: As discussed, don't use classes

File: cypress/integration/login/signup_spec.js [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+        cy.get('#valid_name').should('contain', 'Username must begin with a letter, and contain between 3 to 22 lowercase characters made up of numbers, letters, and the symbols \'.\', \'-\' and \'_\'');
+
+        cy.get('#password_label').should('contain', 'Choose your password');
+        cy.get('#password').should('be.visible');
+
+        cy.get('#createAccountButton').should('be.visible');
+        cy.get('#createAccountButton').should('contain', 'Create Account');
+
+        cy.get('#signup_agreement').should('contain', 'By proceeding to create your account and use Mattermost, you agree to our Terms of Service and Privacy Policy. If you do not agree, you cannot use Mattermost.');
+        cy.get('#signup_agreement > span > [href="https://about.mattermost.com/default-terms/"]').should('be.visible');
```
Comment (saturninoabril) [inline cypress/integration/login/signup_spec.js]: I think this is fine as this span is generated by the `FormattedMarkdownMessage` and there's no way to add ID to the link generated by it.  I don't think there'll be future changes on how the `FormattedMarkdownMessage` works.

**Comment by saturninoabril** [comment]
Thanks @jwilander @enahum, please take a look again.

**Pull request merged**